swaybar/i3bar.c

```diff
@@ -125,9 +125,12 @@ bool i3bar_handle_readable(struct status_line *status) {
 			status_error(status, "[failed to allocate buffer]");
 			return -1;
 		}
+		state->current_node += new_buffer - state->buffer;
+		cur += new_buffer - state->buffer;
 		state->buffer = new_buffer;
 	}
 
+	cur[n] = '\0';
```

Comment:
Should read at most `state->buffer_size - state->buffer_index - 1` bytes at line 114?

Reply:
Also, not related to your patch but `read` returns -1 on error.

Reply:
I think if read() ends up reading enough to fill the entire buffer, realloc would get called, and would provide the space for this null terminator.
The logic is not very obvious, though.